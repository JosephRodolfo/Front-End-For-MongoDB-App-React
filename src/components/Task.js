import React from "react";
import moment from "moment";
import { useState } from "react";

const Task = ({taskDeleteParent, taskEditParent, index, content}) => {
  const [edited, setEdit] = useState(true);

  const handleEditClick = () => {
    setEdit(!edited);
  };

  const handleSave = (e) => {
    e.preventDefault();
    let number = e.currentTarget.id;
    const { completed = false, description = "" } = document.forms.namedItem(number);
    setEdit(!edited);

    const body = {
    completed: completed.checked || false,
    description: description.value || "",
    };

    taskEditParent(content._id, body);
  };
  return (
    <div className="task-card-container">
      <div>
        {edited ? (
          <div>
            <div className="task-desc-checkbox">
              <p>Description: {content.description}</p>

              <div className="edit-task-group">
                <span>Completed: </span>
                <input
                  type="checkbox"
                  readOnly
                  checked={content.completed}
                />
              </div>
            </div>
            <button onClick={handleEditClick}>Edit</button>
          </div>
        ) : (
          <form role="form" id={index.toString()} onSubmit={handleSave}>
            <div className="task-desc-checkbox">
              <div className="edit-task-group">
                <label htmlFor="description">Description: </label>
                <input name="description" className="text-input" type="text" />
              </div>

              <div className="edit-task-group">
                <label htmlFor="completed">Completed: </label>
                <input
                  type="checkbox"
                  name="completed"
                  defaultChecked={content.completed}
                />
              </div>
            </div>
            <button type="submit" id={content._id}>
              Save
            </button>
          </form>
        )}
      </div>

      <p>{moment(content.createdAt).format("MMMM Do, YYYY")}</p>
      <button
        id={content._id}
        onClick={(e) => {
          let taskToDelete = e.target.id;
          taskDeleteParent(taskToDelete);
        }}
      >
        Delete
      </button>
    </div>
  );
};

export default Task;
