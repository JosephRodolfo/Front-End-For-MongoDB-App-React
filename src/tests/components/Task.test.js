import React from "react";
import Task from "../../components/Task";
import "@testing-library/jest-dom";
import ShallowRenderer from "react-test-renderer/shallow";
import { content } from "../fixtures/content";
import { render, screen, fireEvent } from "@testing-library/react";
let result;
let taskEditParent;
let taskDeleteParent;
beforeEach(() => {
  taskEditParent = jest.fn();
  taskDeleteParent = jest.fn();
  const renderer = new ShallowRenderer();
  renderer.render(<Task content={content} />);
  result = renderer.getRenderOutput();
});

test("should render Task correctly", () => {
  expect(result.type).toBe("div");
  expect(result).toMatchSnapshot();
});

test("should conditionally render form for editing Task correctly when clicking edit button", () => {
  const index = 0;
  render(<Task index={index} content={content} />);

  fireEvent.click(screen.getByText("Edit"));

  expect(screen.getByRole("form")).toBeInTheDocument();
});

test("should handle clicking save correctly after editing task", () => {
  const index = 1;
  render(
    <Task index={index} taskEditParent={taskEditParent} content={content} />
  );

  fireEvent.click(screen.getByText("Edit"));
  fireEvent.click(screen.getByText("Save"));

  expect(taskEditParent).toHaveBeenCalled();
  expect(taskEditParent).toHaveBeenCalledWith(content._id, {
    completed: false,
    description: "",
  });
});

test("should handle clicking delete task correctly", () => {
  const index = 1;
  render(
    <Task
      index={index}
      taskDeleteParent={taskDeleteParent}
      taskEditParent={taskEditParent}
      content={content}
    />
  );

  fireEvent.click(screen.getByText("Delete"));

  expect(taskDeleteParent).toHaveBeenCalled();
  expect(taskDeleteParent).toHaveBeenCalledWith(content._id);
});
