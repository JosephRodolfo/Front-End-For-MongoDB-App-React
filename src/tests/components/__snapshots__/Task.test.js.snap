// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Task correctly 1`] = `
<div
  className="task-card-container"
>
  <div>
    <div>
      <div
        className="task-desc-checkbox"
      >
        <p>
          Description: 
          test text
        </p>
        <div
          className="edit-task-group"
        >
          <span>
            Completed: 
          </span>
          <input
            checked={false}
            readOnly={true}
            type="checkbox"
          />
        </div>
      </div>
      <button
        onClick={[Function]}
      >
        Edit
      </button>
    </div>
  </div>
  <p>
    May 4th, 2022
  </p>
  <button
    id="FixtureNonSenseID"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
